### ISSUE TITLE
Flavors don't seem to work with bare_metal

### ISSUE THREAD

**Issue by causton81** [fog/fog-softlayer#50]
I am getting “Unable to find prices for block device 0 with capacity of 250” when trying to create bare-metal with `m1.small` flavor.
 
excon.error.request                                                                                                                                                                  [23/3953]
  :body                => "{\"parameters\":[{\"datacenter\":{\"name\":\"sjc01\"},\"hostname\":\"chef\",\"domain\":\"fog.sldev.us\",\"hourlyBillingFlag\":true,\"operatingSystemReferenceCode\"
:\"CENTOS_6_64\",\"primaryNetworkComponent\":null,\"primaryBackendNetworkComponent\":null,\"sshKeys\":[{\"id\":77986}],\"privateNetworkOnlyFlag\":false,\"networkComponents\":[{\"maxSpeed\":1
0}],\"processorCoreAmount\":2,\"memoryCapacity\":2,\"hardDrives\":[{\"capacity\":250}]}]}"
  :chunk_size          => 1048576
  :ciphers             => "HIGH:!SSLv2:!aNULL:!eNULL:!3DES"
  :connect_timeout     => 60
  :debug_request       => true
  :debug_response      => true
  :expects             => [
    200
    201
  ]
  :headers             => {
    "Authorization"  => "REDACTED"
    "Content-Length" => 398
    "Content-Type"   => "application/json"
    "Host"           => "api.softlayer.com:443"
    "User-Agent"     => "Fog SoftLayer Adapter 0.4.0"
  }
  :host                => "api.softlayer.com"
  :hostname            => "api.softlayer.com"
  :idempotent          => false
  :instrumentor_name   => "excon"
  :method              => :POST
  :middlewares         => [
    Excon::Middleware::ResponseParser
    Excon::Middleware::Expects
    Excon::Middleware::Idempotent
    Excon::Middleware::Instrumentor
    Excon::Middleware::Mock
  ]
  :mock                => false
  :nonblock            => true
  :omit_default_port   => false
  :password            => "REDACTED"
  :path                => "/rest/v3/SoftLayer_Hardware_Server/createObject.json"
  :persistent          => false
  :port                => 443
  :query               => nil
  :read_timeout        => 60
  :retries_remaining   => 4
  :retry_limit         => 4
  :scheme              => "https"
  :ssl_verify_peer     => true
  :tcp_nodelay         => false
  :thread_safe_sockets => true
  :uri_parser          => URI
  :user                => "chris.auston"
  :versions            => "excon/0.44.1 (x86_64-linux) ruby/2.1.4"
  :instrumentor_name   => "excon"                                                                                                                                                     [0/3953]
  :method              => :POST
  :middlewares         => [
    Excon::Middleware::ResponseParser
    Excon::Middleware::Expects
    Excon::Middleware::Idempotent
    Excon::Middleware::Instrumentor
    Excon::Middleware::Mock
  ]
  :mock                => false
  :nonblock            => true
  :omit_default_port   => false
  :password            => "REDACTED"
  :path                => "/rest/v3/SoftLayer_Hardware_Server/createObject.json"
  :persistent          => false
  :port                => 443
  :query               => nil
  :read_timeout        => 60
  :retries_remaining   => 4
  :retry_limit         => 4
  :scheme              => "https"
  :ssl_verify_peer     => true
  :tcp_nodelay         => false
  :thread_safe_sockets => true
  :uri_parser          => URI
  :user                => "chris.auston"
  :versions            => "excon/0.44.1 (x86_64-linux) ruby/2.1.4"
  :write_timeout       => 60
excon.error.response
  :body          => "{\"error\":\"Unable to find prices for block device 0 with capacity of 250.\",\"code\":\"SoftLayer_Exception_NotFound\"}"
  :headers       => {
    "Connection"           => "close"
    "Content-Length"       => "112"
    "Content-Type"         => "application/json"
    "Date"                 => "Mon, 23 Mar 2015 15:27:22 GMT"
    "Server"               => "Apache"
    "Vary"                 => "Accept-Encoding"
    "X-Backside-Transport" => "FAIL FAIL"
    "X-Client-IP"          => "192.168.67.115"
  }
  :local_address => "172.30.198.119"
  :local_port    => 58142
  :reason_phrase => "Internal Server Error"
  :remote_ip     => "66.228.119.116"
  :status        => 500

**Comment by mfbmina** [comment]
Hi @causton81 ! It's a problem with flavor!
Taking a look on [bare metal create options](https://github.com/fog/fog-softlayer/blob/master/lib/fog/softlayer/requests/compute/get_bare_metal_create_options.rb#L233) we can see that there isn't a option for 250 GB, only for 500 GB.

Try create a BM without flavor or  wait for a fix on flavors! :smile:

**Comment by causton81** [comment]
Thanks, I was able to place an order with explicit :cpu, :ram, and :disk
options.

On Fri, Mar 27, 2015 at 2:42 PM, Matheus Mina <notifications@github.com>
wrote:

> Hi @causton81 <https://github.com/causton81> ! It's a problem with flavor!
> Taking a look on bare metal create options
> <https://github.com/fog/fog-softlayer/blob/master/lib/fog/softlayer/requests/compute/get_bare_metal_create_options.rb#L233>
> we can see that there isn't a option for 250 GB, only for 500 GB.
>
> Try create a BM without flavor or wait for a fix on flavors! [image:
> :smile:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fog/fog-softlayer/issues/50#issuecomment-87068371>.
>



-- 
-ca

**Comment by mfbmina** [comment]
@cphrmky For close this issue, the only thing is change the example for create a BM? If there isn't anything else, I can do that!

**Comment by causton81** [comment]
My 2 cents: it would be nice if SLAPI supported BM flavors as a first-class concept. Then I can write a script (or chef-provisioning) saying I want a m1.tiny BM.  Ideally that m1.tiny maps to a quick provisioning package so I don't have to wait a long time. Also if the hardware specs change for "m1.tiny" my script should still work.  If the specs change the price may too, but AFAIK fog-softlayer already requires the user to figure that out on their own?

**Comment by fernandes** [comment]
hi @causton81 
I updated your comment to use GFM and be easier to read.

Really would be great if SLAPI have flavors on its API, but as we are not related to SL (except @cphrmky) we can't say anything about it heheheh We just can fix fog using the current API

I'm merging @mfbmina PR, cut a new release then you can try it

**Comment by fernandes** [comment]
@causton81 could you please post how you're trying to buy this server, @mfbmina just changed the example on his PR, I do really believe we need to change something related to [flavors](https://github.com/fog/fog-softlayer/blob/master/lib/fog/softlayer/models/compute/flavors.rb#L21) after you post how you're doing I dig into it

thank you

**Comment by causton81** [comment]
Full disclosure, I work for SoftLayer too on the Object Storage team ;)

Below are the smallest parameters I was able to use to make a successful order:

cpu.servers.create(:bare_metal => true, :cpu=>2, :ram => 2, :datacenter=>'tor01', :name=>'fogbmtest', :os_code => 'DEBIAN_7_64', :disk => [{:capacity => 500}])

**Comment by fernandes** [comment]
hahahah so you have inside information :stuck_out_tongue_winking_eye: 

now you pointed me the right direction, on fog we are using the same flavors for BM and CCI, so a refactor is needed to use specific flavors for each case, I'm gonna leave it opened and assigned to me, so I can work on this when I have some bandwidth

thank you :)

**Comment by causton81** [comment]
Thank you for contributing! I have been using fog-softlayer with chef-provisioning-fog (and chef-provisioning-fog-softlayer which is only on my github at the moment) and it works very well with VSIs. I'd love to get the BM side working too so we can support all kinds of development pipelines.

**Comment by fernandes** [comment]
uow.. thats great

so, could you please provide me values that makes sense for BM `m1.tiny`, `m1.small`, `m1.medium`, `m1.large`, `m1.xlarge`

cpu / disk / ram

thank you!

**Comment by causton81** [comment]
This looks like the existing logic which seems to force 250 or 500 for first disk on BM:
  https://github.com/fog/fog-softlayer/blob/master/lib/fog/softlayer/models/compute/server.rb#L453-L457

I can test different combinations but I'm not really sure what to tell you.  Maybe @cphrmky has an opinion?

**Comment by fernandes** [comment]
makes sense... I'm gonna make some tests then I reply here... ty!

**Comment by mfbmina** [comment]
I agree with you @fernandes, we need have specific flavors for each server type. For BM, the minimum disk value is 500 GB
